compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -494,8 +497,23 @@ impl<'p, 'tcx> MatchVisitor<'_, 'p, 'tcx> {
                         " else { todo!() }",
                         Applicability::HasPlaceholders,
                     );
+                } else {
+                    // If the pattern to match is an integer literal, perhaps the user meant to create a binding instead.
+                    if let PatKind::Lit(expr) = pat.kind {
+                        if let ExprKind::Lit(Spanned { node: LitKind::Int(_, _), span }) = expr.kind
+                        {
+                            err.span_suggestion_verbose(
+                                span.shrink_to_lo(),
+                                "alternatively, prepend the pattern with an underscore to define \
+                                      a variable; identifiers cannot start with digits",
+                                "_",
+                                Applicability::MaybeIncorrect,
+                            );
```

Comment:
Open to this, but it seems to me that I haven't really wanted to define an integer-named variable ~ever. Maybe instead of a help: text with `_int` we can just `note: ` that identifiers cannot start with digits? I think in 99% of cases they probably wanted to use if let or similar, so adding a second suggestion may be more confusing.

At least to me if I see two suggestions I think I'd think "which is better?" before thinking "oh, I wanted an if, the second one is irrelevant".

Reply:
Obviously this is somewhat targeted at new developers but, personally, I've never pattern matched on a integer literal before. I'd use a match statement instead which allows better handling of the default case, etc. FWIW, I think that if a user tries to do this (again, IMO very uncommon) then really meant to assign a value instead.

Also I think PR will need a rebase because of #106458.